scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -541,6 +556,163 @@ def _get_containers(client, ignore_container=None, restrict_to_container=None, l
+                        continue
+
+                    short_cid = _get_short_cid( cid )
+
+                    # ignore the scalyr-agent container
+                    if cid == self._container_id:
+                        global_log.log( scalyr_logging.DEBUG_LEVEL_2, "Excluding container because cid '%s' is the scalyr_agent container" % short_cid )
+                        continue
+
+                    # make sure the container has a name
+                    cname = container.get( 'name' )
+                    if cname is None:
+                        global_log.log( scalyr_logging.DEBUG_LEVEL_2, "Excluding container '%s' because no container name was found" % short_cid)
+                        continue
+
+                    # we are only interested in running or terminated pods
+                    cstate = container.get( 'state', {} )
+                    if 'running' not in cstate and 'terminated' not in cstate:
+                        global_log.log( scalyr_logging.DEBUG_LEVEL_2, "Excluding container '%s' because not yet running" % short_cid )
+                        continue
+
+                    add_container = True
+
+                    # filter against the container glob list
+                    if glob_list:
+                        add_container = False
+                        for glob in glob_list:
+                            if fnmatch.fnmatch( cname, glob ):
+                                add_container = True
+                                break
+
+                    if not add_container:
+                        global_log.log( scalyr_logging.DEBUG_LEVEL_2, "Excluding container '%s' based on glob_list" % short_cid)
+                        continue
+
+                    # build a k8s_info structure similar to the one built by _get_containers
+                    k8s_info = {}
+                    k8s_info['pod_name'] = pod_name
+                    k8s_info['pod_namespace'] = pod_namespace
+                    k8s_info['pod_uid'] = pod_uid
+                    k8s_info['k8s_container_name'] = cname
+
+                    # get pod and deployment/controller information for the container
+                    if k8s_cache:
+                        pod = k8s_cache.pod( pod_namespace, pod_name, current_time )
+                        if pod:
+                            # check to see if we should exclude this container
+                            default_exclude = not k8s_include_by_default
+                            exclude = pod.exclude_pod( container_name=cname, default=default_exclude)
+
+                            # exclude if necessary
+                            if exclude:
+                                if pod.annotations:
+                                    global_log.log( scalyr_logging.DEBUG_LEVEL_2, "Excluding container '%s' based on pod annotations, %s" % (short_cid, str(pod.annotations)) )
```

Comment:
> doesn't the %s implicitly run a str() on whatever you pass?

Hmm, I'm sure i've encountered parts of the code where it was necessary to put a str around some objects - specifically when printing exception objects, and you'll see that's where the majority of this usage happens in the code base - though maybe it's just something I picked up from the existing codebase, because you're right, according to the docs and it appears python does call `str` when `%s` is used, and I did a quick test and didn't notice any problems.  I wonder if it's something that was relevant to older versions?